samba: add python dependencies to buildInputs

amaxine opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
```
package = pkgs.samba4.override {
  enableLDAP = true;
  enablePrinting = true;
  enableMDNS = true;
  enableDomainController = true;
  enableRegedit = true;
};
```

I couldn't build Samba after the most recent bump (#124693), this fixed things. It seems both are only needed for AD, I haven't tested building it in other configurations than above.

I should probably trim my configuration - I don't think I actually need ldap support at all here :sweat_smile: - but I thought this would be worth PRing regardless. There might be more breakage elsewhere, and I imagine this needs backporting same as the other PR.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Commit:
samba: add missing python dependencies for ldap and domain controller

r-rmcgibbo commented:
Result of `nixpkgs-review pr 125105` at c8eeabbf run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/6d35141a50219029786ddbb60fda91ae">sambaFull</a> (samba4Full)</li>
  </ul>
</details>
<details>
  <summary>5 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:58:5:
```
   |
58 |     ./patch-source3__libads__kerberos_keytab.c.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/servers/samba/4.x.nix:157:5:
```
    |
157 |     license = licenses.gpl3;
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:59:5:
```
   |
59 |     ./4.x-no-persistent-install-dynconfig.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:57:5:
```
   |
57 |     ./4.x-no-persistent-install.patch
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/servers/samba/4.x.nix:60:5:
```
   |
60 |     ./4.x-fix-makeflags-parsing.patch
   |     ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

mweinelt committed:
samba4Full: disable glusterfs support

The samba package was marked as broken, when enableGlusterFS is true.

The samba build with glusterfs fails due to API breakage that I am
unable to debug:

[3562/4088] Compiling source3/modules/vfs_virusfilter.c
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_pread’:
../../source3/modules/vfs_glusterfs.c:856:8: error: too few arguments to function ‘glfs_pread’
  856 |  ret = glfs_pread(glfd, data, n, offset, 0);
      |        ^~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:713:1: note: declared here
  713 | glfs_pread(glfs_fd_t *fd, void *buf, size_t count, off_t offset, int flags,
      | ^~~~~~~~~~
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_pread_do’:
../../source3/modules/vfs_glusterfs.c:938:16: error: too few arguments to function ‘glfs_pread’
  938 |   state->ret = glfs_pread(state->fd, state->buf, state->count,
      |                ^~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:713:1: note: declared here
  713 | glfs_pread(glfs_fd_t *fd, void *buf, size_t count, off_t offset, int flags,
      | ^~~~~~~~~~
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_pwrite_do’:
../../source3/modules/vfs_glusterfs.c:1077:16: error: too few arguments to function ‘glfs_pwrite’
 1077 |   state->ret = glfs_pwrite(state->fd, state->buf, state->count,
      |                ^~~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:717:1: note: declared here
  717 | glfs_pwrite(glfs_fd_t *fd, const void *buf, size_t count, off_t offset,
      | ^~~~~~~~~~~
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_pwrite’:
../../source3/modules/vfs_glusterfs.c:1161:8: error: too few arguments to function ‘glfs_pwrite’
 1161 |  ret = glfs_pwrite(glfd, data, n, offset, 0);
      |        ^~~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:717:1: note: declared here
  717 | glfs_pwrite(glfs_fd_t *fd, const void *buf, size_t count, off_t offset,
      | ^~~~~~~~~~~
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_fsync_do’:
../../source3/modules/vfs_glusterfs.c:1287:16: error: too few arguments to function ‘glfs_fsync’
 1287 |   state->ret = glfs_fsync(state->fd);
      |                ^~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:790:1: note: declared here
  790 | glfs_fsync(glfs_fd_t *fd, struct glfs_stat *prestat,
      | ^~~~~~~~~~
../../source3/modules/vfs_glusterfs.c: In function ‘vfs_gluster_ftruncate’:
../../source3/modules/vfs_glusterfs.c:1621:8: error: too few arguments to function ‘glfs_ftruncate’
 1621 |  ret = glfs_ftruncate(glfd, offset);
      |        ^~~~~~~~~~~~~~
In file included from ../../source3/modules/vfs_glusterfs.c:41:
/nix/store/0gzaf6fqgfxfns19zlc07dyjqigj7ak7-glusterfs-9.0/include/glusterfs/api/glfs.h:768:1: note: declared here
  768 | glfs_ftruncate(glfs_fd_t *fd, off_t length, struct glfs_stat *prestat,
      | ^~~~~~~~~~~~~~

../../source3/modules/vfs_virusfilter.c: In function ‘quarantine_create_dir’:
../../source3/modules/vfs_virusfilter.c:132:13: warning: implicit declaration of function ‘strlcat’; did you mean ‘strncat’? [-Wimplicit-function-declaration]
  132 |   cat_len = strlcat(new_dir, "/", len + 1);
      |             ^~~~~~~
      |             strncat

Waf: Leaving directory `/build/samba-4.14.4/bin/default'
Build failed
 -> task in 'vfs_glusterfs.objlist' failed with exit status 1 (run with -v to display more information)

mweinelt commented:
Hi, sorry for being so pushy, this is also broken on release-21.05 and I'm in a hurry to get that fixed.

The package still failed to build, so here is what I changed:

- Move `python3Packages.dnspython` over into the domain controller section, it is used in a dnsupdate script, not part of LDAP
- Fixed LDAP by including `openldap.dev` output, instead ouf the default `openldap.out`
- Finally remove `enableGlusterFS` from `samba4Full` to unbreak the package

maxeaubrey commented:
No worries. I placed `dnspython` under LDAP based on docs more than anything, but I could have just misread those. Thanks for fixing the rest up, including the full package!

SuperSandro2000 commented on pkgs/servers/samba/4.x.nix:
```diff
@@ -91,10 +91,10 @@ stdenv.mkDerivation rec {
     tdb
   ] ++ optionals stdenv.isLinux [ liburing systemd ]
-    ++ optional enableLDAP openldap
+    ++ optional enableLDAP [ openldap.dev python3Packages.markdown ]
```
```suggestion
    ++ optionals enableLDAP [ openldap.dev python3Packages.markdown ]
```

mweinelt replied:
Thanks, fixed.
